pkg/kv/kvserver/multiqueue/multi_queue.go

```diff
@@ -0,0 +1,231 @@
+	}
+	queue.lastQueueIndex = -1
+
+	return &queue
+}
+
+// Permit is a token which is returned from a Task.GetWaitChan call.
+type Permit struct {
+	valid bool
+}
+
+// tryRunNextLocked will run the next task in order round-robin through the
+// queues and in priority order within a queue. It will return true if it ran a
+// task. The MultiQueue.mu lock must be held before calling this func.
+func (m *MultiQueue) tryRunNextLocked() {
+	// If no permits are left, then we can't run anything.
+	if m.remainingRuns <= 0 {
+		return
+	}
+
+	for i := 0; i < len(m.outstanding); i++ {
+		// Start with the next queue in order and iterate through all empty queues.
+		// If all queues are empty then return false signaling that nothing was run.
+		index := (m.lastQueueIndex + i + 1) % len(m.outstanding)
+		if m.outstanding[index].Len() > 0 {
+			task := heap.Pop(&m.outstanding[index]).(*Task)
+			task.permitC <- &Permit{valid: true}
+			m.remainingRuns--
+			m.lastQueueIndex = index
+			return
+		}
+	}
+}
+
+// Add returns a Task that must be closed (calling Task.Close) to
+// release the Permit. The number of types is expected to
+// be relatively small and not be changing over time.
+func (m *MultiQueue) Add(queueType int, priority float64) *Task {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// The mutex starts locked, unlock it when we are ready to run.
+	pos, ok := m.mapping[queueType]
+	if !ok {
+		// Append a new entry to both mapping and outstanding each time there is
+		// a new queue type.
+		pos = len(m.outstanding)
+		m.mapping[queueType] = pos
+		m.outstanding = append(m.outstanding, notifyHeap{})
+	}
+	newTask := Task{
+		priority:  priority,
+		permitC:   make(chan *Permit, 1),
+		heapIdx:   -1,
+		queueType: queueType,
+	}
+	heap.Push(&m.outstanding[pos], &newTask)
+
+	// Once we are done adding a task, signal the main loop in case it finished
+	// all its work and was waiting for more work. We are holding the mu lock when
+	// signaling, so we guarantee that it will not be able to respond to the
+	// signal until after we release the lock.
+	m.tryRunNextLocked()
+
+	return &newTask
+}
+
+// Cancel will cancel a Task that may not have started yet. This is useful if it
+// is determined that it is no longer required to run this Task.
+func (m *MultiQueue) Cancel(task *Task) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// Find the right queue and try to remove it. Queues monotonically grow, and a
+	// Task will track its position within the queue.
+	queueIdx := m.mapping[task.queueType]
+	ok := m.outstanding[queueIdx].tryRemove(task)
+	// If we get here, we are racing with the task being started. The concern is
+	// that the caller may also call MultiQueue.Release since the task was
+	// started. Either we get the permit or the caller, so we guarantee only one
+	// release will be called.
+	if !ok {
+		select {
+		case p, ok := <-task.permitC:
+			// Only release if the channel is open, and we can get the permit.
+			if ok {
+				p.valid = false
```

Comment:
Consider extracting a `releaseLocked` method that you can call from here and from `Release` to avoid repetition. Something like:
```go
func (m *MultiQueue) Release(permit *Permit) {
	if !permit.valid {
		panic("double release of permit")
	}
	m.mu.Lock()
	defer m.mu.Unlock()
  m.releaseLocked()
}

func (m *MultiQueue) releaseLocked(permit *Permit) {
  p.valid = false
  m.remainingRuns++
	m.tryRunNextLocked()
}
```